improvement: mark commits with autodaved content when starting a new environment

lorenzo-cavazzi opened the pull request:
Implement the solution suggested in #1113 to make it clear to the users which commits have autosaved content.

***Preview:*** https://lorenzotest.dev.renku.ch/
***How to test:*** Start an interactive environment on any project, modify a file and stop the environment from the environments page in the UI. Try to start a new envireonment and verify that the relevant commit is marked with a `*`. Select it to get a short explanation.

![Screenshot_20201118_152646](https://user-images.githubusercontent.com/43481553/99544513-af7d2b00-29b4-11eb-9863-ee1e1322e0c2.png)

![Screenshot_20201118_152707](https://user-images.githubusercontent.com/43481553/99544489-a7bd8680-29b4-11eb-9034-73fdb2fca7db.png)

fix #1113

ciyer commented on client/src/notebooks/Notebooks.present.js:
```diff
@@ -1055,19 +1055,41 @@ class StartNotebookPipelinesContent extends Component {
+        commitComment = (
+          <FormText>
+            <FontAwesomeIcon className="no-pointer" icon={faInfoCircle} /> For this commit we
+            found <ExternalLink url={url} iconSup={true} iconAfter={true} title="autosaved content" role="link" />
```
Suggested wording:
`We found <ExternalLink url={url} iconSup={true} iconAfter={true} title="unsaved work" role="link" /> for this commit.`

ciyer commented on client/src/notebooks/Notebooks.present.js:
```diff
@@ -1433,6 +1457,10 @@ class AutosavedDataModal extends Component {
             We will automatically restore this content so you do not lose any work.
           </p>
+          <p>
+            If you don&apos;t need it, you can restore the using the following command:
```
Text: `If you do not need it, you can discard this work with the following command:`

ciyer commented on client/src/notebooks/Notebooks.present.js:
```diff
@@ -1433,6 +1457,10 @@ class AutosavedDataModal extends Component {
           </p>
+          <p>
+            If you don&apos;t need it, you can restore the using the following command:
+            <br /><code>git reset --hard {this.props.filters.commit.short_id} && git clean -f -d</code>
```
Could we wrap the command in a `<Clipboard>[...]</Clipboard>` to make it easy to copy?

ciyer commented on client/src/notebooks/Notebooks.present.js:
```diff
@@ -1415,14 +1438,15 @@ class AutosavedDataModal extends Component {
+    );
+    const docsLink = (
+      <ExternalLink
+        role="text" iconSup={true} iconAfter={true} title="documentation page"
```
Title can be just `documentation` (instead of "documentation page").

ciyer reviewed (changes requested):
This is nice. Even though the '*' does not have any documentation, I think users will quickly understand what it means (they will pick a commit with a '*' to see why that is there, then they will see the explanation that there is unsaved work here).

I have some suggested changes to the wording and a request to make it easier to copy the text. Could we also change the title of the pop-up to: `Unsaved work`?

<img width="717" alt="image" src="https://user-images.githubusercontent.com/1196411/99684370-aefe9600-2a81-11eb-9a89-666aa8c88c71.png">

BTW, is it possible to indicate on the environment listing that this commit was started from an autosave branch? And could we show the command there as well?

lorenzo-cavazzi commented:
I addressed all the comments apart from this:

> BTW, is it possible to indicate on the environment listing that this commit was started from an autosave branch? And could we show the command there as well?

As discussed offline, that's trickier because we don't get that information from renku-notebooks. We determine the availability of autosaved content based on the branch names, and we get the branches directly from GitLab. Since the branch is deleted as soon as the autosaved content is restored, we don't have that information anymore. We could try to save that in the UI but a simple UI refresh would permanently remove it.
Also, the user can get rid of the autosaved content in the JupyterLab environment and the UI would still mention it. This may be confusing.

Preview updated here: https://lorenzotest.dev.renku.ch/

![Screenshot_20201120_121722](https://user-images.githubusercontent.com/43481553/99794418-6bac3200-2b2a-11eb-8178-c31b3ffef510.png)

ciyer reviewed (approved):
Very nice!

rokroskar reviewed (approved):
This works as advertised! The only thing I'm wondering is whether there should be some indication that _some_ autosave branches exist at the top level of the interactive environment page - otherwise the user doesn't really know that there is some saved state until they look at the commit listing. We can discuss that separately though.

lorenzo-cavazzi commented:
That's a good point. On one hand, it may be useful information, but on the other hand, it may not be relevant to the user to know that there is autosaved content _at some previous commit_. If there are more recent commits, the old autosaved content is very probably outdated.

The one reason why this consideration may not apply is that we don't associate the autosaved content to the latest commit synchronized with the remote. We can expect users to keep working on environments started from outdated commits, ending up with an autosave for an old commit.

We can continue the discussion (and possibly add a mockup) here: #1143

rokroskar commented:
> The one reason why this consideration may not apply is that we don't associate the autosaved content to the latest commit synchronized with the remote.

yep, exactly. This is the sort of situation I got into when I was testing this PR and I wasn't sure at all how it should be resolved.
